File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/UserHandler.java
package com.yahyeet.boardbook.model.handler;

import com.yahyeet.boardbook.model.entity.Match;
import com.yahyeet.boardbook.model.entity.User;
import com.yahyeet.boardbook.model.handler.populator.MatchPlayerPopulator;
import com.yahyeet.boardbook.model.handler.populator.MatchPopulator;
import com.yahyeet.boardbook.model.handler.populator.UserPopulator;
import com.yahyeet.boardbook.model.repository.IGameRepository;
import com.yahyeet.boardbook.model.repository.IGameRoleRepository;
import com.yahyeet.boardbook.model.repository.IGameTeamRepository;
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository;
import com.yahyeet.boardbook.model.repository.IMatchRepository;
import com.yahyeet.boardbook.model.repository.IRepositoryListener;
import com.yahyeet.boardbook.model.repository.IUserRepository;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

/**
 * A class that handles user entities
 */
public class UserHandler implements IRepositoryListener<User>, IEntityHandler<User> {
	private IUserRepository userRepository;
	private List<IUserHandlerListener> listeners = new ArrayList<>();

	private MatchPlayerPopulator matchPlayerPopulator;
	private MatchPopulator matchPopulator;
	private UserPopulator userPopulator;

	public UserHandler(IUserRepository userRepository,
										 IMatchRepository matchRepository,
										 IGameRoleRepository gameRoleRepository,
										 IGameTeamRepository gameTeamRepository,
										 IGameRepository gameRepository,
										 IMatchPlayerRepository matchPlayerRepository) {
		this.userRepository = userRepository;
		matchPlayerPopulator = new MatchPlayerPopulator(gameRoleRepository, gameTeamRepository, matchRepository, userRepository);
		matchPopulator = new MatchPopulator(gameRepository, matchPlayerRepository);
		userPopulator = new UserPopulator(matchPlayerRepository, matchRepository, userRepository);

		this.userRepository.addListener(this);
	}

	@Override
	public CompletableFuture<User> find(String id) {
		return find(id, generatePopulatorConfig(false, false));
	}

	@Override
	public CompletableFuture<User> find(String id, Map<String, Boolean> config) {
		return userRepository.find(id).thenCompose(user -> populate(user, config));
	}

	@Override
	public CompletableFuture<User> save(User user) {
		return save(user, generatePopulatorConfig(false, false));
	}

	@Override
	public CompletableFuture<User> save(User user, Map<String, Boolean> config) {
		checkUserValidity(user);

		return userRepository.save(user).thenCompose(savedUser -> populate(savedUser, config));
	}

	@Override
	public CompletableFuture<List<User>> all() {
		return all(generatePopulatorConfig(false, false));
	}

	@Override
	public CompletableFuture<List<User>> all(Map<String, Boolean> config) {
		return userRepository.all().thenComposeAsync(users -> {
			List<CompletableFuture<User>> completableFutures = users
				.stream()
				.map(user -> populate(user, config))
				.collect(Collectors.toList());

			CompletableFuture<Void> allFutures = CompletableFuture.allOf(
				completableFutures
					.toArray(new CompletableFuture[0])
			);
			return allFutures.thenApplyAsync(
				future -> completableFutures
					.stream()
					.map(CompletableFuture::join)
					.collect(Collectors.toList()));
		});
	}

	private CompletableFuture<User> populate(User user, Map<String, Boolean> config) {
		AtomicReference<User> fullyPopulatedUser = new AtomicReference<>();

		return userPopulator.populate(user, config).thenApply(populatedUser -> {
			fullyPopulatedUser.set(populatedUser);

			return null;
		}).thenCompose(nothing -> {
			List<CompletableFuture<Match>> allFuturePopulatedMatches =
				fullyPopulatedUser
					.get()
					.getMatches()
					.stream().map(match -> matchPopulator.populate(match, MatchHandler.generatePopulatorConfig(true, true)))
					.collect(Collectors.toList());

			CompletableFuture<Void> allOfFutureFuturePopulatedMatches = CompletableFuture.allOf(allFuturePopulatedMatches.toArray(new CompletableFuture[0]));

			return allOfFutureFuturePopulatedMatches.thenApply(future ->
				allFuturePopulatedMatches
					.stream()
					.map(CompletableFuture::join)
					.collect(Collectors.toList()));
		}).thenApply(populatedMatches -> {
			fullyPopulatedUser.get().getMatches().clear();
			populatedMatches.forEach(populatedMatch -> fullyPopulatedUser.get().addMatch(populatedMatch));

			return populatedMatches;
		}).thenCompose(populatedMatches -> {
			List<CompletableFuture<Void>> allFuturePopulatedMatchPlayers = new ArrayList<>();

			fullyPopulatedUser
				.get()
				.getMatches()
				.forEach(match -> {
					match
						.getMatchPlayers()
						.forEach(matchPlayer -> allFuturePopulatedMatchPlayers.add(matchPlayerPopulator.populate(matchPlayer).thenApply(populatedMatchPlayer -> {
							matchPlayer.setUser(populatedMatchPlayer.getUser());
							matchPlayer.setMatch(populatedMatchPlayer.getMatch());
							matchPlayer.setTeam(populatedMatchPlayer.getTeam());
							matchPlayer.setRole(populatedMatchPlayer.getRole());

							return null;
						})));
				});

			return CompletableFuture.allOf(allFuturePopulatedMatchPlayers.toArray(new CompletableFuture[0]));
		}).thenApply(nothing -> fullyPopulatedUser.get());
	}

	public void addListener(IUserHandlerListener listener) {
		listeners.add(listener);
	}

	public void removeListener(IUserHandlerListener listener) {
		listeners.remove(listener);
	}

	private void notifyListenersOnUserAdd(User user) {
		for (IUserHandlerListener listener : listeners) {
			listener.onAddUser(user);
		}
	}

	private void notifyListenersOnUserUpdate(User user) {
		for (IUserHandlerListener listener : listeners) {
			listener.onUpdateUser(user);
		}
	}

	private void notifyListenersOnUserRemove(String id) {
		for (IUserHandlerListener listener : listeners) {
			listener.onRemoveUser(id);
		}
	}

	@Override
	public void onCreate(User user) {
		notifyListenersOnUserAdd(user);
	}

	@Override
	public void onUpdate(User user) {
		notifyListenersOnUserUpdate(user);
	}

	@Override
	public void onDelete(String id) {
		notifyListenersOnUserRemove(id);
	}

	public static Map<String, Boolean> generatePopulatorConfig(
		boolean shouldPopulateFriends,
		boolean shouldPopulateMatches
	) {
		Map<String, Boolean> config = new HashMap<>();
		config.put("friends", shouldPopulateFriends);
		config.put("matches", shouldPopulateMatches);
		return config;
	}

	/**
	 * Checks whether a user contains valid data or not
	 *
	 * @param user User to be validated
	 */
	private void checkUserValidity(User user) {
		for (User friend : user.getFriends()) {
			if (friend.getId() == null) {
				throw new IllegalArgumentException("Cannot add friends that have no id");
			}
		}
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/handler/UserHandler.kt
package com.yahyeet.boardbook.model.handler

import com.yahyeet.boardbook.model.entity.Game
import com.yahyeet.boardbook.model.entity.Match
import com.yahyeet.boardbook.model.entity.User
import com.yahyeet.boardbook.model.handler.populator.MatchPlayerPopulator
import com.yahyeet.boardbook.model.handler.populator.MatchPopulator
import com.yahyeet.boardbook.model.handler.populator.UserPopulator
import com.yahyeet.boardbook.model.repository.IGameRepository
import com.yahyeet.boardbook.model.repository.IGameRoleRepository
import com.yahyeet.boardbook.model.repository.IGameTeamRepository
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository
import com.yahyeet.boardbook.model.repository.IMatchRepository
import com.yahyeet.boardbook.model.repository.IRepositoryListener
import com.yahyeet.boardbook.model.repository.IUserRepository

import java.util.ArrayList
import java.util.concurrent.CompletableFuture
import java.util.concurrent.atomic.AtomicReference
import java.util.stream.Collectors

class UserHandler(private val userRepository: IUserRepository,
                  matchRepository: IMatchRepository,
                  gameRoleRepository: IGameRoleRepository,
                  gameTeamRepository: IGameTeamRepository,
                  gameRepository: IGameRepository,
                  matchPlayerRepository: IMatchPlayerRepository) : IRepositoryListener<User>, EntityHandler<User> {
    private val listeners = ArrayList<UserHandlerListener>()

    private val matchPlayerPopulator: MatchPlayerPopulator
    private val matchPopulator: MatchPopulator
    private val userPopulator: UserPopulator

    init {
        matchPlayerPopulator = MatchPlayerPopulator(gameRoleRepository, gameTeamRepository, matchRepository, userRepository)
        matchPopulator = MatchPopulator(gameRepository, matchPlayerRepository)
        userPopulator = UserPopulator(matchPlayerRepository, matchRepository, userRepository)
    }

    override fun find(id: String): CompletableFuture<User> {
        return userRepository.find(id).thenCompose(Function<User, CompletionStage<User>> { this.populate(it) })
    }

    override fun save(user: User): CompletableFuture<User> {
        checkUserValidity(user)

        return userRepository.save(user).thenCompose(Function<User, CompletionStage<User>> { this.populate(it) })
    }

    override fun all(): CompletableFuture<List<User>> {
        return userRepository.all().thenComposeAsync { users ->
            val completableFutures = users
                    .stream()
                    .map<CompletableFuture<User>>(Function<User, CompletableFuture<User>> { this.populate(it) })
                    .collect<List<CompletableFuture<User>>, Any>(Collectors.toList())

            val allFutures = CompletableFuture.allOf(
                    *completableFutures
                            .toTypedArray<CompletableFuture>()
            )
            allFutures.thenApplyAsync { future ->
                completableFutures
                        .stream()
                        .map<User>(Function<CompletableFuture<User>, User> { it.join() })
                        .collect<List<User>, Any>(Collectors.toList())
            }
        }
    }

    fun populate(user: User): CompletableFuture<User> {
        val fullyPopulatedUser = AtomicReference<User>()

        return userPopulator.populate(user).thenApply<Any> { populatedUser ->
            fullyPopulatedUser.set(populatedUser)
            null
        }.thenCompose { nothing ->
            val allFuturePopulatedMatches = fullyPopulatedUser
                    .get()
                    .matches
                    .stream().map { match -> matchPopulator.populate(match) }
                    .collect<List<CompletableFuture<Match>>, Any>(Collectors.toList())

            val allOfFutureFuturePopulatedMatches = CompletableFuture.allOf(*allFuturePopulatedMatches.toTypedArray<CompletableFuture>())

            allOfFutureFuturePopulatedMatches.thenApply { future ->
                allFuturePopulatedMatches
                        .stream()
                        .map<Match>(Function<CompletableFuture<Match>, Match> { it.join() })
                        .collect<List<Match>, Any>(Collectors.toList())
            }
        }.thenApply { populatedMatches ->
            fullyPopulatedUser.get().matches.clear()
            populatedMatches.forEach { populatedMatch -> fullyPopulatedUser.get().addMatch(populatedMatch) }

            populatedMatches
        }.thenCompose { populatedMatches ->
            val allFuturePopulatedMatchPlayers = ArrayList<CompletableFuture<Void>>()

            fullyPopulatedUser
                    .get()
                    .matches
                    .forEach { match ->
                        match
                                .matchPlayers
                                .forEach { matchPlayer ->
                                    allFuturePopulatedMatchPlayers.add(matchPlayerPopulator.populate(matchPlayer).thenApply { populatedMatchPlayer ->
                                        matchPlayer.user = populatedMatchPlayer.user
                                        matchPlayer.match = populatedMatchPlayer.match
                                        matchPlayer.role = populatedMatchPlayer.role
                                        null
                                    })
                                }
                    }

            CompletableFuture.allOf(*allFuturePopulatedMatchPlayers.toTypedArray<CompletableFuture>())
        }.thenApply { nothing -> fullyPopulatedUser.get() }
    }

    fun addListener(listener: UserHandlerListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: UserHandlerListener) {
        listeners.remove(listener)
    }

    private fun notifyListenersOnUserAdd(user: User) {
        for (listener in listeners) {
            listener.onAddUser(user)
        }
    }

    private fun notifyListenersOnUserUpdate(user: User) {
        for (listener in listeners) {
            listener.onUpdateUser(user)
        }
    }

    private fun notifyListenersOnUserRemove(user: User) {
        for (listener in listeners) {
            listener.onRemoveUser(user)
        }
    }

    override fun onCreate(user: User) {
        notifyListenersOnUserAdd(user)
    }

    override fun onUpdate(user: User) {
        notifyListenersOnUserUpdate(user)
    }

    override fun onDelete(user: User) {
        notifyListenersOnUserRemove(user)
    }

    private fun checkUserValidity(user: User) {
        for (friend in user.friends) {
            kotlin.requireNotNull(friend.id) { "Cannot add friends that have no id" }
        }
    }
}
